src/main/java/org/truffleruby/core/range/RubyRange.java

```diff
@@ -9,17 +9,37 @@
  */
 package org.truffleruby.core.range;
 
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
 import com.oracle.truffle.api.object.Shape;
 import org.truffleruby.core.klass.RubyClass;
 import org.truffleruby.language.RubyDynamicObject;
+import org.truffleruby.language.library.RubyLibrary;
 
+@ExportLibrary(RubyLibrary.class)
 public abstract class RubyRange extends RubyDynamicObject {
 
     public boolean excludedEnd;
+    public boolean frozen;
 
     public RubyRange(RubyClass rubyClass, Shape shape, boolean excludedEnd) {
         super(rubyClass, shape);
         this.excludedEnd = excludedEnd;
     }
```

Comment:
In general I prefer to keep a small number of constructors, this makes the data flow clearer (fewer things access the fields directly) and also show what value callers give for every argument.
I can do that change (removing the old constructor, or at least delegating to the new one).

Reply:
Actually, could you remove the old constructor in this PR to avoid any conflict?
An easy way to do it is to have the old constructor call the new one, and then use Refactor->Inlined Method.

I guess you'll do the change so that Range instances are created frozen in the next PR?

Reply:
Yes I recommended a new constructor, as the vast majority of uses of the existing constructor don't need to change.

Reply:
Actually they all need to change once we implement the new semantics, isn't it? And having the new constructor makes it clear each caller needs to decide the appropriate value for frozen.

I think this would be clearer if it was together with the semantics change.
I don't mind separate PRs but anyway here we need at minimum that the old constructor would delegate to the new one for keeping the code clean and understandable (e.g., it's pretty subtle why the default is `false` here for frozen).

Given the usages, I guess changing to the old constructor delegating to the new one seems the easiest fix, and we'll update transitive callers and remove the old constructor in the next PR then?

Reply:
> Actually they all need to change once we implement the new semantics, isn't it?

No most will take the default value.

But we can simplify this discussion by removing the new constructor, and only talking about it in the concrete context of the new PR.

Reply:
Another option I didn't think before is the new constructor is not needed in this PR. Anyway I think we all agree it will be clearer with the other changes :)